src/manage_nsfs/nc_lifecycle.js

```diff
@@ -2,44 +2,141 @@
 'use strict';
 
 const dbg = require('../util/debug_module')(__filename);
+const os = require('os');
 const _ = require('lodash');
+const path = require('path');
 const util = require('util');
 const P = require('../util/promise');
 const config = require('../../config');
 const nb_native = require('../util/nb_native');
 const NsfsObjectSDK = require('../sdk/nsfs_object_sdk');
+const native_fs_utils = require('../util/native_fs_utils');
+const { NoobaaEvent } = require('./manage_nsfs_events_utils');
 const ManageCLIError = require('./manage_nsfs_cli_errors').ManageCLIError;
-const { throw_cli_error, get_service_status, NOOBAA_SERVICE_NAME } = require('./manage_nsfs_cli_utils');
+const ManageCLIResponse = require('../manage_nsfs/manage_nsfs_cli_responses').ManageCLIResponse;
+const { write_stdout_response, throw_cli_error, get_service_status, NOOBAA_SERVICE_NAME } = require('./manage_nsfs_cli_utils');
 
-// TODO: 
+// TODO:
 // implement 
 // 1. notifications
 // 2. POSIX scanning and filtering per rule
 // 3. GPFS ILM policy and apply for scanning and filtering optimization
 
+const CLUSTER_LOCK = 'cluster.lock';
+const config_fs_options = { silent_if_missing: true };
+
+/**
+ * lifecycle_timeout returns a promise that rejects with a timeout error
+ * @returns {Promise}
+ */
+async function lifecycle_timeout() {
+    // TODO - change to manage_nsfs_error?
+    return new Promise((resolve, reject) => setTimeout(() => reject(new Error('lifecycle worker reached timeout')), config.NC_LIFECYCLE_TIMEOUT_MS));
+}
+
+/**
+ * run_lifecycle_under_lock runs the lifecycle workflow under a file system lock
+ * lifecycle workflow is being locked to prevent multiple instances from running the lifecycle workflow
+ * @param {import('../sdk/config_fs').ConfigFS} config_fs 
+ */
+async function run_lifecycle_under_lock(config_fs) {
+    const lifecyle_logs_dir_path = config.LIFECYCLE_LOGS_DIR;
+    await config_fs.create_dir_if_missing(lifecyle_logs_dir_path);
+    const lock_path = path.join(lifecyle_logs_dir_path, CLUSTER_LOCK);
+
+    await native_fs_utils.lock_and_run(config_fs.fs_context, lock_path, async () => {
```

Comment:
this function got a bit messy. I think you should probably move the lambda function to be a proper function with relevant name